controller.py

```diff
@@ -1,75 +1,165 @@
-
     # Allocate output variable
     # For this example, only thrustSI, and torque members are relevant
     control = cffirmware.control_t()
-
     # allocate desired state
     setpoint = cffirmware.setpoint_t()
-    # select trajectory tracking mode
+    setpoint = setTrajmode(setpoint)
+    sensors = cffirmware.sensorData_t()
+    state = cffirmware.state_t()
+    return control, setpoint, sensors, state 
+
+def setTrajmode(setpoint):
     setpoint.mode.x = cffirmware.modeAbs
     setpoint.mode.y = cffirmware.modeAbs
     setpoint.mode.z = cffirmware.modeAbs
     setpoint.mode.roll = cffirmware.modeDisable
     setpoint.mode.pitch = cffirmware.modeDisable
     setpoint.mode.yaw = cffirmware.modeDisable
+    return setpoint
 
-    # allocate current state (both sensor and state are the state)
-    # This is kind of odd and should be part of state
-    sensors = cffirmware.sensorData_t()
-    state = cffirmware.state_t()
+def updateDesState(setpoint, fulltraj):
+    # fulltraj[3::] = 0
+    setpoint.position.x = fulltraj[0]  # m
+    setpoint.position.y = fulltraj[1]  # m
+    setpoint.position.z = fulltraj[2]  # m
+    setpoint.velocity.x = fulltraj[3]  # m/s
+    setpoint.velocity.y = fulltraj[4]  # m/s
+    setpoint.velocity.z = fulltraj[5]  # m/s
+    setpoint.acceleration.x = fulltraj[6]  # m/s^2
+    setpoint.acceleration.y = fulltraj[7]  # m/s^2
+    setpoint.acceleration.z = fulltraj[8]  # m/s^2
+    # setpoint.attitude.yaw = 0  # deg
+    return setpoint
+    
+def updateSensor(sensors,uavState):
+    sensors.gyro.x = np.degrees(uavState[10]) # deg/s
+    sensors.gyro.y = np.degrees(uavState[11]) # deg/s
+    sensors.gyro.z = np.degrees(uavState[12]) # deg/s
+    return sensors
+
+
+def updateState(state, uavState):
+    state.position.x = uavState[0]   # m
+    state.position.y = uavState[1]    # m
+    state.position.z = uavState[2]    # m
+    state.velocity.x = uavState[3]    # m/s
+    state.velocity.y = uavState[4]    # m/s
+    state.velocity.z = uavState[5]    # m/s
+    q_curr = np.array(uavState[6:10]).reshape((4,))
+    rpy_state  = to_euler(q_curr,convention='xyz')
+    state.attitude.roll  = np.degrees(rpy_state[0])
+    state.attitude.pitch = np.degrees(-rpy_state[1])
+    state.attitude.yaw   = np.degrees(rpy_state[2])
+    fullState = np.array([state.position.x,state.position.y,state.position.z, 
+                          state.velocity.x,state.velocity.y, state.velocity.z, 
+                          q_curr[0],q_curr[1],q_curr[2],q_curr[3], 0,0,0]).reshape((13,))
+    return state,fullState
 
+def stateUavStateError(state, uavState):    
+    state_ = np.array([state.position.x, state.position.y, state.position.z, state.attitude.roll, state.attitude.pitch, state.attitude.yaw])
+    q_curr = np.array(uavState[6:10]).reshape((4,))
+    rpy_state  = to_euler(q_curr,convention='xyz')
+    uavState_ = np.array([uavState[0], uavState[1], uavState[2],rpy_state[0], rpy_state[1], rpy_state[2]])
+    return state_ - uavState_
```

Comment:
For rotation, this is not a proper distance metric. Rowan has some better metrics for orientation error. Where is this function used?

Reply:
This was just debugging to check that what is passed from the dynamic model states to the controller is the same, I forgot to delete it because it was just for checking but it is misrepresented.